App crashes on start if location services are not turned on

E ActivityManager(749)        App crashed! Process: org.azavea.otm2.dev
E AndroidRuntime(6132)        FATAL EXCEPTION: main
E AndroidRuntime(6132)        Process: org.azavea.otm2.dev, PID: 6132
E AndroidRuntime(6132)        java.lang.RuntimeException: Unable to resume activity {org.azavea.otm2.dev/org.azavea.otm.ui.InstanceSwitcherActivity}: java.lang.NullPointerException
E AndroidRuntime(6132)          at android.app.ActivityThread.performResumeActivity(ActivityThread.java:3321)
E AndroidRuntime(6132)          at android.app.ActivityThread.handleResumeActivity(ActivityThread.java:3359)
E AndroidRuntime(6132)          at android.app.ActivityThread$H.handleMessage(ActivityThread.java:1397)
E AndroidRuntime(6132)          at android.os.Handler.dispatchMessage(Handler.java:102)
E AndroidRuntime(6132)          at android.os.Looper.loop(Looper.java:157)
E AndroidRuntime(6132)          at android.app.ActivityThread.main(ActivityThread.java:5872)
E AndroidRuntime(6132)          at java.lang.reflect.Method.invokeNative(Native Method)
E AndroidRuntime(6132)          at java.lang.reflect.Method.invoke(Method.java:515)
E AndroidRuntime(6132)          at com.android.internal.os.ZygoteInit$MethodAndArgsCaller.run(ZygoteInit.java:858)
E AndroidRuntime(6132)          at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:674)
E AndroidRuntime(6132)          at dalvik.system.NativeStart.main(Native Method)
E AndroidRuntime(6132)        Caused by: java.lang.NullPointerException
E AndroidRuntime(6132)          at org.azavea.otm.ui.InstanceSwitcherActivity.onStart(InstanceSwitcherActivity.java:162)
E AndroidRuntime(6132)          at android.app.Instrumentation.callActivityOnStart(Instrumentation.java:1239)
E AndroidRuntime(6132)          at android.app.Activity.performStart(Activity.java:5322)
E AndroidRuntime(6132)          at android.app.Activity.performRestart(Activity.java:5378)
E AndroidRuntime(6132)          at android.app.Activity.performResume(Activity.java:5397)
E AndroidRuntime(6132)          at android.app.ActivityThread.performResumeActivity(ActivityThread.java:3304)
E AndroidRuntime(6132)          ... 10 more

--- a/OpenTreeMap/res/layout/instance_switcher_activity.xml
+++ b/OpenTreeMap/res/layout/instance_switcher_activity.xml
@@ -16,8 +16,14 @@
 
     <TextView
         android:id="@+id/instance_list_empty"
-        style="@style/instance_switcher_label"
+        style="@style/instance_switcher_empty_list_view"
         android:text="@string/instance_switcher_no_instances_found"
+        android:visibility="gone" />
+
+    <TextView
+        android:id="@+id/instance_list_location_off"
+        style="@style/instance_switcher_empty_list_view"
+        android:text="@string/instance_switcher_location_off_text"
         android:visibility="gone" />
 
     <TextView

--- a/OpenTreeMap/res/values/strings.xml
+++ b/OpenTreeMap/res/values/strings.xml
@@ -135,5 +135,6 @@
     <string name="filter_instances_hint">Filter tree maps…</string>
     <string name="instances_failed">Could not get tree maps</string>
     <string name="instance_switcher_public_instances">See All Public Tree Maps</string>
+    <string name="instance_switcher_location_off_text">Location services must be turned on in order to see any nearby tree maps.</string>
 
 </resources>

--- a/OpenTreeMap/res/values/styles.xml
+++ b/OpenTreeMap/res/values/styles.xml
@@ -143,6 +143,12 @@
         <item name="android:layout_height">wrap_content</item>
     </style>
 
+    <style name="instance_switcher_empty_list_view" parent="@style/instance_switcher_label">
+        <item name="android:layout_width">match_parent</item>
+        <item name="android:layout_height">0dp</item>
+        <item name="android:layout_weight">1</item>
+    </style>
+
     <style name="add_tree_label">
         <item name="android:background">@color/main_background_color</item>
         <item name="android:gravity">center</item>

--- a/OpenTreeMap/src/org/azavea/otm/ui/InstanceSwitcherActivity.java
+++ b/OpenTreeMap/src/org/azavea/otm/ui/InstanceSwitcherActivity.java
@@ -44,6 +44,8 @@ public class InstanceSwitcherActivity extends Activity {
     private User user = null;
     private Location userLocation;
 
+    LocationManager locationManager;
+
     private ProgressDialog loadingInstances;
     private ProgressDialog loadingInstance;
 
@@ -51,6 +53,18 @@ public class InstanceSwitcherActivity extends Activity {
     public void onCreate(Bundle savedInstanceState) {
         super.onCreate(savedInstanceState);
         setContentView(R.layout.instance_switcher_activity);
+
+        findViewById(R.id.login_button).setOnClickListener(v -> startActivity(new Intent(InstanceSwitcherActivity.this, LoginActivity.class)));
+        findViewById(R.id.logout_button).setOnClickListener(v -> {
+            App.getLoginManager().logOut(InstanceSwitcherActivity.this);
+            startActivity(new Intent(InstanceSwitcherActivity.this, LoginActivity.class));
+        });
+        findViewById(R.id.public_instances_button).setOnClickListener(v ->
+                startActivityForResult(
+                        new Intent(InstanceSwitcherActivity.this, PublicInstanceListDisplay.class),
+                        INSTANCE_SELECT_REQUEST_CODE));
+
+        locationManager = (LocationManager) App.getAppInstance().getSystemService(Context.LOCATION_SERVICE);
     }
 
     @Override
@@ -118,11 +132,8 @@ public class InstanceSwitcherActivity extends Activity {
     }
 
     // TODO: backport this algorithm, it's more fault tolerant
-    private static Location getBestLocation(Criteria accuracyCrit) {
+    private Location getBestLocation(Criteria accuracyCrit) {
         Location location = null;
-        Context context = App.getAppInstance();
-        LocationManager locationManager =
-                (LocationManager) context.getSystemService(Context.LOCATION_SERVICE);
 
         if (locationManager != null) {
             final String bestProvider = locationManager.getBestProvider(accuracyCrit, true);
@@ -149,12 +160,15 @@ public class InstanceSwitcherActivity extends Activity {
         super.onStart();
 
         // Only setup instance lists if the logged in user has changed or moved from their location
-        Criteria criteria = new Criteria();
+        final Criteria criteria = new Criteria();
         criteria.setAccuracy(Criteria.ACCURACY_COARSE);
-        Location newLocation = getBestLocation(criteria);
-        User newUser = App.getLoginManager().loggedInUser;
+        final Location newLocation = getBestLocation(criteria);
+        final User newUser = App.getLoginManager().loggedInUser;
 
-        if (user != newUser || areLocationsDistant(userLocation, newLocation)) {
+        if (newLocation == null) {
+            final ListView instancesView = (ListView) findViewById(R.id.instance_list);
+            instancesView.setEmptyView(findViewById(R.id.instance_list_location_off));
+        } else if (user != newUser || areLocationsDistant(userLocation, newLocation)) {
             RequestGenerator rg = new RequestGenerator();
 
             loadingInstances = ProgressDialog.show(this, getString(R.string.instance_switcher_dialog_heading),
@@ -168,16 +182,6 @@ public class InstanceSwitcherActivity extends Activity {
 
         userLocation = newLocation;
         user = newUser;
-
-        findViewById(R.id.login_button).setOnClickListener(v -> startActivity(new Intent(InstanceSwitcherActivity.this, LoginActivity.class)));
-        findViewById(R.id.logout_button).setOnClickListener(v -> {
-            App.getLoginManager().logOut(InstanceSwitcherActivity.this);
-            startActivity(new Intent(InstanceSwitcherActivity.this, LoginActivity.class));
-        });
-        findViewById(R.id.public_instances_button).setOnClickListener(v ->
-                startActivityForResult(
-                        new Intent(InstanceSwitcherActivity.this, PublicInstanceListDisplay.class),
-                        INSTANCE_SELECT_REQUEST_CODE));
     }
 
     /**

